Add an AWS S3 scan result storage

carlos-ardila2 opened the pull request:
Add support for using AWS S3 as online cache for scan results.

Provides a S3FileStorage class which extends [FileStorage](https://github.com/oss-review-toolkit/ort/blob/main/utils/ort/src/main/kotlin/storage/FileStorage.kt) and uses official Amazon's  [AWS Java SDK](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/examples-s3.html) to store scan results.

Issue oss-review-toolkit/ort#730

tsteenbe commented on model/src/main/kotlin/config/S3FileStorageConfiguration.kt:
```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (C) 2017 The ORT Project Authors (see <https://github.com/oss-review-toolkit/ort/blob/main/NOTICE>)
```
S3FileStorageConfiguration.kt is a new file so please change copyright year from 2017 to 2023

carlos-ardila2 replied:
Addressed by squashing (not the best way to do it, I agree)

tsteenbe commented on utils/ort/src/funTest/kotlin/storage/S3FileStorageFunTest.kt:
```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (C) 2019 The ORT Project Authors (see <https://github.com/oss-review-toolkit/ort/blob/main/NOTICE>)
```
S3FileStorageFunTest.kt is a new file so please change copyright year from 2019 to 2023

carlos-ardila2 replied:
Addressed by squashing

tsteenbe commented on utils/ort/src/main/kotlin/storage/S3FileStorage.kt:
```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright (C) 2019 The ORT Project Authors (see <https://github.com/oss-review-toolkit/ort/blob/main/NOTICE>)
```
S3FileStorage.kt is a new file so please change copyright year from 2019 to 2023

carlos-ardila2 replied:
Addressed by squashing

tsteenbe commented on README.md:
```diff
@@ -546,6 +546,30 @@ ort:
 
+### AWS S3 Storage
+
+An AWS S3 Bucket can be used to store scan results. You must provide a previously created Bucket, an optional AWS region
```
Could you make "AWS S3 Bucket" a link to https://docs.aws.amazon.com/AmazonS3/latest/userguide/Welcome.html? Will help ORT users quickly find the docs on how to set up a bucket.

tsteenbe commented on README.md:
```diff
@@ -546,6 +546,30 @@ ort:
+ort:
+  scanner:
+    storages:
+      awsStorage:
```
AWS offers multiple storage services see https://aws.amazon.com/products/storage/ therefore I believe it's better to not name the key `awsStorage` in case we want to add other AWS storage service in the future.

What do you think of renaming `awsStorage` to `awsS3Storage` across this PR?

sschuberth commented on gradle/libs.versions.toml:
```diff
@@ -42,12 +42,13 @@ logback = "1.4.11"
 mavenResolver = "1.9.15"
 mockk = "1.13.7"
-okhttp = "4.11.0"
+okhttp = "5.0.0-alpha.11"
```
Why's this necessary? We shouldn't use alpha version dependencies in production code (and ORT's `main` *is* used in production, as it's supposed to always be stable).

carlos-ardila2 replied:
Amazon has [stated](https://github.com/awslabs/aws-sdk-kotlin/issues/765#issuecomment-1374093175) that their Koltin library relies on the latest version of okhttp, which oddly enough is the alpha one.  I'm not sure how to proceed.

sschuberth replied:
> which oddly enough is the alpha one.

Thanks for this information, indeed odd. I don't think it's acceptable for us to depend on an alpha version of OkHttp for the whole of ORT. We could solve this if 1) [scan storages were plugin](https://github.com/oss-review-toolkit/ort/issues/6603) and 2) [plugin classpaths were isolated from ORT classpath](https://github.com/oss-review-toolkit/ort/issues/6798), but both are bigger refactorings that I do not see happening any time soon.

> I'm not sure how to proceed.

I guess our best option then is to indeed use the Java instead of the Kotlin flavor of the AWS SDK...

carlos-ardila2 replied:
I'll start working on a refactor.

carlos-ardila2 replied:
I have completed the refactor, please review.

carlos-ardila2 replied:
I'm sorry, I'm afraid I need some clarification. There are two commits on the branch right now, the first one contains the version that uses the experimental Kotlin library by Amazon. Then another commit with a refactor switching to Java library, therefore, this alpha version (and other changes mentioned below) are found **solely** in the first commit but **not** in the HEAD. Should this first commit be squashed too? What I usually do is to keep commits until just before merging, then I squash them, just in case I need to track down errors and even discarded ideas. Could you please explain me this project's policy regarding that?

sschuberth replied:
> There are two commits on the branch right now, the first one contains the version that uses the experimental Kotlin library by Amazon.

Ah, my bad, I might have accidentally looked only at the first commit, assuming there would be only one / believing to look at the combined diff.

> Should this first commit be squashed too?

Yes please. We don't squash upon merge (as that combines individual commits from a PR), but rebase upon merge (also merge commit are not allowed).

> What I usually do is to keep commits until just before merging, then I squash them, just in case I need to track down errors and even discarded ideas.

Makes sense, but locally you could probably use `git reflog` to restore intermediate states.

carlos-ardila2 replied:
Got it! Thans for the `git reflog ` tip, life saver.

sschuberth commented on gradle/libs.versions.toml:
```diff
@@ -42,12 +42,13 @@ logback = "1.4.11"
 reflections = "0.10.2"
 retrofit = "2.9.0"
 retrofitConverterKotlinxSerialization = "1.0.0"
+s3 = "0.29.0-beta"
```
Also here, please use the latest stable / non-beta version, if possible.

carlos-ardila2 replied:
[0.32.4-beta](https://mvnrepository.com/artifact/aws.sdk.kotlin/s3/0.32.4-beta) is the latest version of the official Kotlin SDK provided by Amazon. We could try to use their Java SDK, [2.20.162](https://mvnrepository.com/artifact/software.amazon.awssdk/s3/2.20.162) But I believe it will defy the purpose of having the best version available from the vendor, which uses actual Kotlin features.

sschuberth replied:
Ok, I see, the AWS SDK for Kotlin is a developer preview, and there hasn't been any non-beta release since version [0.16.0](https://mvnrepository.com/artifact/aws.sdk.kotlin/s3/0.16.0) from 2022. So let's stick to the latest beta version then.

carlos-ardila2 replied:
Using the latest now. Although not important now that we are refactoring to use the java library

sschuberth commented on gradle/libs.versions.toml:
```diff
@@ -70,6 +71,7 @@ versions = { id = "com.github.ben-manes.versions", version.ref = "versionsPlugin
 antlr = { module = "org.antlr:antlr4", version.ref = "antlr" }
 asciidoctorj = { module = "org.asciidoctor:asciidoctorj", version.ref = "asciidoctorj" }
 asciidoctorjPdf = { module = "org.asciidoctor:asciidoctorj-pdf", version.ref = "asciidoctorjPdf" }
+aws-s3 = { module = "aws.sdk.kotlin:s3", version.ref = "s3" }
```
Use CamelCase instead of snake-case for the key (see existing code).

carlos-ardila2 replied:
Fixed, thank you

sschuberth commented on model/src/main/kotlin/config/S3FileStorageConfiguration.kt:
```diff
@@ -0,0 +1,42 @@
+ * A class to hold the configuration for using local files as a storage.
+ */
+data class S3FileStorageConfiguration(
+    /**
+     * The name of the S3 bucket used to store files in.
+     */
```
You may want to use more compact one-liners like

    /** The name of the S3 bucket used to store files in. */

to document properties.

Also, consider adding empty lines to separate the properties, so it's visually more clear which comment belongs to which property.

carlos-ardila2 replied:
Addressed. Although, there are many other files in the codebase that still use three lines of comment per field.

sschuberth commented on model/src/test/kotlin/config/OrtConfigurationTest.kt:
```diff
@@ -229,6 +231,7 @@ class OrtConfigurationTest : WordSpec({
                         this shouldContainExactly mapOf(
                             "commandLine" to "--copyright --license --info --strip-root --timeout 300",
                             "commandLineNonConfig" to "--processes 4",
+                            "parseLicenseExpressions" to "true",
```
This should be dropped.

carlos-ardila2 replied:
Addressed by squashing.

sschuberth commented on utils/ort/src/funTest/kotlin/storage/S3FileStorageFunTest.kt:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    init {
+
```
Please drop this superfluous empty line.

carlos-ardila2 replied:
Addressed

sschuberth commented on utils/ort/src/main/kotlin/OkHttpClientHelper.kt:
```diff
@@ -209,8 +209,8 @@ fun OkHttpClient.download(url: String, acceptEncoding: String? = null): Result<P
 
-        if (!response.isSuccessful || body == null) {
-            body?.close()
+        if (!response.isSuccessful) {
+            body.close()
```
Why this change? Does AWS allow null bodies on success? If so, this should be extracted to a preceding preparing refactoring commit, that thoroughly explains this in the commit message.

sschuberth replied:
Ah, wait, this is due to the OkHttp alpha upgrade, and thus should be dropped together with the upgrade.

carlos-ardila2 replied:
Fixed with refactor.

carlos-ardila2 replied:
Only in the first commit (won't compile otherwise), gone in the HEAD.

sschuberth commented on utils/ort/src/main/kotlin/storage/HttpFileStorage.kt:
```diff
@@ -105,12 +105,9 @@ class HttpFileStorage(
         val response = httpClient.execute(request)
         if (response.isSuccessful) {
-            response.body?.let { body ->
+            response.body.let { body ->
```
Drop this, see above.

carlos-ardila2 replied:
Fixed with refactor.

carlos-ardila2 replied:
Only in the first commit (won't compile otherwise), gone in the HEAD.

sschuberth commented on utils/ort/src/main/kotlin/storage/S3FileStorage.kt:
```diff
@@ -0,0 +1,175 @@
+ */
+@Suppress("SwallowedException")
+class S3FileStorage(
+    /**
+     * The name of the S3 bucket used to store files in.
+     */
```
Consider more compact comments / empty separator lines as outlined above.

carlos-ardila2 replied:
Addressed

sschuberth commented on utils/ort/src/main/kotlin/storage/S3FileStorage.kt:
```diff
@@ -0,0 +1,175 @@
+
+        var fileExists = true
+
+        runBlocking {
+            try {
+                s3.headObject(request)
+            } catch (e: NotFound) {
+                fileExists = false
+            }
+        }
```
Probably better something like

    return runBlocking { runCatching { s3.headObject(request) }.isSuccess }

carlos-ardila2 replied:
Thank you for that valuable suggestion. I'd still like to keep some record if the expection is any different form NotFound.

sschuberth commented on utils/ort/src/main/kotlin/storage/S3FileStorage.kt:
```diff
@@ -0,0 +1,175 @@
+            }
+        }
+
+        val stream: InputStream
+
+        runBlocking {
+            stream = job.await()
+        }
+
+        return stream
```
`return runBlocking { job.await() }`

carlos-ardila2 replied:
No longer part of the code base after refactor.

sschuberth commented on utils/ort/src/main/kotlin/storage/S3FileStorage.kt:
```diff
@@ -0,0 +1,175 @@
+                try {
+                    s3.putObject(it)
+                } catch (e: S3Exception) {
+                    logger.warn("Can not write $path to S3 bucket $bucketName")
```
Use lazy curly-brace syntax with `log`-statements: `log.warn {}`

carlos-ardila2 replied:
Addressed

sschuberth commented on utils/ort/src/main/kotlin/storage/S3FileStorage.kt:
```diff
@@ -0,0 +1,175 @@
+            }
+        }
+
+        // Local server doesn't play nice with auto scheduled coroutines. Run blocking instead
```
Please properly end (even comment) sentences with a dot, and generally use proper grammar like in a written user's manual, not informal language.

carlos-ardila2 replied:
Addressed

sschuberth commented on model/src/main/kotlin/config/S3FileStorageConfiguration.kt:
```diff
@@ -30,13 +30,17 @@ data class S3FileStorageConfiguration(
      * The AWS secret for the access key.
      */
-     val secretAccessKey: String?
+    val secretAccessKey: String?,
```
These lines should be amended to the previous commit.

carlos-ardila2 replied:
Addressed by squashing.

sschuberth commented on README.md:
```diff
@@ -551,7 +551,7 @@ ort:
 (if not present it will use [default](https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/region-selection.html)) 
 and credentials (if not provided, system [default](https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/credential-providers.html) 
-values will be used).
+values will be used). Optionally, you can set the data to be stored compressed.
```
The commit message should probably also mention that compression is optional.

carlos-ardila2 replied:
Addressed by squashing.

sschuberth commented on README.md:
```diff
@@ -548,16 +548,18 @@ ort:
-(if not present it will use [default](https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/region-selection.html))
-and credentials (if not provided, system [default](https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/credential-providers.html)
-values will be used). Optionally, you can set the data to be stored compressed.
+An [AWS S3 Bucket](https://docs.aws.amazon.com/AmazonS3/latest/userguide/Welcome.html) can be used to store scan
```
Again, move the hunks in this commit to the commits that introduced the original respective code.

carlos-ardila2 replied:
Addressed by squashing

sschuberth commented:
@carlos-ardila2 is there any interest from your side to move this forward?

carlos-ardila2 commented on integrations/schemas/ort-configuration-schema.json:
```diff
@@ -540,6 +540,7 @@
         },
         "StorageTypes": {
             "enum": [
+                "aws",
```
Addressed by squashing

carlos-ardila2 commented on README.md:
```diff
@@ -385,7 +385,7 @@ supported:
     [no peer dependencies](https://github.com/oss-review-toolkit/ort/issues/95))
   * [Yarn 1](https://classic.yarnpkg.com/)
-  * [Yarn 2+](https://next.yarnpkg.com/)
+  * [Yarn 2+](https://yarnpkg.com/)
```
Addressed by squashing.

carlos-ardila2 replied:
Addressed by squashing.

carlos-ardila2 commented:
@sschuberth I'm back in. I apologize for taking this long to get back to this task. The AWS requirement to use an alpha version for the Kotlin S3 library was a set-back. I've addressed most of the issues and squash in a single commit. I know it is not the best way to do it, but I thought it would be best to start fresh for the refactor. Regards.

carlos-ardila2 commented:
@sschuberth I'm back in. I apologize for taking this long to get back to this task. The AWS requirement to use an alpha version for the Kotlin S3 library was a set-back. I've addressed most of the issues and squash in a single commit. I know it is not the best way to do it, but I thought it would be best to start fresh for the refactor. Regards.

sschuberth commented on utils/ort/src/main/kotlin/storage/S3FileStorage.kt:
```diff
@@ -0,0 +1,173 @@
+        }
+
+        return runBlocking { job.await() }
+
```
Please drop this trailing blank line.

carlos-ardila2 commented:
> @carlos-ardila2 please rebase this once more onto latest `main` to get rid of some unrelated code changes ("afterwards" -> "afterward") and to address unrelated test failures.

Done

sschuberth commented on README.md:
```diff
@@ -546,6 +546,32 @@ ort:
+  scanner:
+    storages:
+      awsS3Storage:
+        backend:
```
Ping @carlos-ardila2, please see above.

sschuberth commented on utils/ort/src/funTest/kotlin/storage/S3FileStorageFunTest.kt:
```diff
@@ -0,0 +1,152 @@
+import org.apache.logging.log4j.kotlin.logger
+
+class S3FileStorageFunTest : WordSpec() {
+
```
Nit: Please drop this superfluous blank line.

sschuberth commented on utils/ort/src/funTest/kotlin/storage/S3FileStorageFunTest.kt:
```diff
@@ -0,0 +1,152 @@
+    private val bucket = "ort-scan-results"
+
+    private val handler = object : HttpHandler {
+        val requests = Collections.synchronizedMap(mutableMapOf<String, String>())
```
As tests within a single spec are not run in parallel, can we stick to a simple `mutableMapOf<String, String>()` like in `HttpFileStorageFunTest`?

sschuberth commented on utils/ort/src/funTest/kotlin/storage/S3FileStorageFunTest.kt:
```diff
@@ -0,0 +1,152 @@
+                }
+                "GET" -> {
+                    val key = exchange.requestURI.toString().removePrefix("/$bucket/")
+                    requests[key].let { data ->
```
There doe snot seem to be a benefit by using `.let` here, so consider using plain `val data = requests[key]` to reduce nesting of following code.

sschuberth commented on utils/ort/src/main/kotlin/storage/S3FileStorage.kt:
```diff
@@ -0,0 +1,146 @@
+ */
+@Suppress("SwallowedException")
+class S3FileStorage(
+
```
Nit: Please drop this superfluous blank line.

sschuberth commented on utils/ort/src/main/kotlin/storage/S3FileStorage.kt:
```diff
@@ -0,0 +1,146 @@
+    private val secretAccessKey: String? = null
+
+) : FileStorage {
+
```
Nit: Please drop this superfluous blank line.

sschuberth commented on utils/ort/src/main/kotlin/storage/S3FileStorage.kt:
```diff
@@ -0,0 +1,146 @@
+
+    /** The AWS secret for the access key. */
+    private val secretAccessKey: String? = null
+
```
Nit: Please drop this superfluous blank line.

sschuberth commented on utils/ort/src/main/kotlin/storage/S3FileStorage.kt:
```diff
@@ -0,0 +1,146 @@
+            null
+        }
+
+        val client = if (awsRegion != null && provider != null) {
```
Just omit the `val client = ` here and use `if` as an expression that returns the value.

sschuberth commented on utils/ort/src/main/kotlin/storage/S3FileStorage.kt:
```diff
@@ -0,0 +1,146 @@
+            }
+        }
+
+        client
```
This can be dropped when using `if` as an expression, see above.

sschuberth commented on utils/ort/src/main/kotlin/storage/S3FileStorage.kt:
```diff
@@ -0,0 +1,146 @@
+
+) : FileStorage {
+
+    val s3: S3Client? by lazy {
```
Let's name the variable `s3Client` for clarity.

sschuberth replied:
Also, can this really ever be null? I acknowledge this comes from a Java platform call that doesn't support explicit non-nullability. But according to the docs the builder should never return `null`, or?

sschuberth commented:
Test failures are unrelated.
